lib/improver/spotdata_new/apply_lapse_rate.py

```diff
@@ -0,0 +1,206 @@
+                self.grid_metadata_identifier that were found on the input
+                cube.
+        """
+        attributes = cube.attributes
+        attributes = {k: v for (k, v) in attributes.items()
+                      if self.grid_metadata_identifier in k}
+        return attributes
+
+    def _compare_attributes(self, attributes, reference_attributes):
+        """
+        Compare keys and values of attributes with reference_attributes.
+        Raise an exception if they do not match exactly; written by default
+        to use Python 3 behaviour, but currently also works with Python 2. When
+        this is no longer required the TypeError exception and resulting action
+        can be removed.
+
+        Args:
+            attributes (dict):
+                A dictionary of attributes.
+            reference_attributes (dict):
+                A reference dictionary of attributes with which to make a
+                comparison.
+        Raises:
+            ValueError: Raised if the metadata extracted is not identical on
+                        all cubes.
+        """
+        try:
+            match = ((attributes.items() & reference_attributes.items()) ==
+                     reference_attributes.items())
+        except TypeError:
+            match = ((attributes.viewitems() &
+                      reference_attributes.viewitems()) ==
+                     reference_attributes.viewitems())
+        if match is not True:
+            raise ValueError('Cubes do not share the metadata identified '
+                             'by the grid_metadata_identifier ({})'.format(
+                                 self.grid_metadata_identifier))
+
+    def check_grid_match(self, cubes):
+        """
+        Uses the provided grid_metadata_identifier to extract and compare
+        attributes on the input cubes. The expectation is that all the metadata
+        identified should match for the cubes to be deemed compatible.
+
+        Args:
+            cubes (list of iris.cube.Cube items):
+                List of cubes for which the attributes should be tested.
+        """
+        # Allow user to bypass cube comparison by setting identifier to None.
+        if self.grid_metadata_identifier is None:
+            return
+
+        reference_attributes = self._get_attributes(cubes[0])
+        for cube in cubes:
+            attributes = self._get_attributes(cube)
+            self._compare_attributes(attributes, reference_attributes)
+
+    def process(self, spot_data_cube, neighbour_cube, gridded_lapse_rate_cube):
+        """
+        Extract lapse rates from the appropriate grid points and apply them to
+        the spot extracted temperatures.
+
+        The calculation is::
+
+         lapse_rate_adjusted_temperatures = temperatures + lapse_rate *
+         vertical_displacement
+
+        Args:
+            spot_data_cube (iris.cube.Cube):
+                The spot data cube of temperatures extracted from the gridded
```

Comment:
Can we be clearer this is temperatures for a list of spot sites, not just one site.  It's a bit hard to read otherwise.

Reply:
added "for the spot data sites" to try and make this clearer.